c3.php

```diff
@@ -38,18 +38,27 @@
 if (!function_exists('__c3_error')) {
     function __c3_error($message)
     {
-        $errorLogFile = defined('C3_CODECOVERAGE_ERROR_LOG_FILE') ?
-            C3_CODECOVERAGE_ERROR_LOG_FILE :
-            C3_CODECOVERAGE_MEDIATE_STORAGE . DIRECTORY_SEPARATOR . 'error.txt';
-        if (is_writable($errorLogFile)) {
-            file_put_contents($errorLogFile, $message);
+        if (defined('C3_CODECOVERAGE_ERROR_LOG_FILE')) {
+            $errorLogFile = C3_CODECOVERAGE_ERROR_LOG_FILE;
+        } elseif (defined('C3_CODECOVERAGE_MEDIATE_STORAGE')) {
+            $errorLogFile = C3_CODECOVERAGE_MEDIATE_STORAGE . DIRECTORY_SEPARATOR . 'error.txt';
         } else {
-            $message = "Could not write error to log file ($errorLogFile), original message: $message";
+            $errorLogFile = null;
         }
+
+        if (isset($errorLogFile)) {
+            if (@file_put_contents($errorLogFile, $message) === false) {
+                $message = "Could not write error to log file ($errorLogFile), original message: $message";
+            }
+        }
+
         if (!headers_sent()) {
             header('X-Codeception-CodeCoverage-Error: ' . str_replace("\n", ' ', $message), true, 500);
         }
         setcookie('CODECEPTION_CODECOVERAGE_ERROR', $message);
+
+        echo 'CodeCeption C3 Error: ' . $message;
+        exit(1);
```

Comment:
This block is actually a major change in behaviour.
Before, c3 set a cookie and website continue to work as usual, now it fails loud and clear and the website doesn't work. It isn't a problem for tests, but, if c3.php is included to live website, it would break website for real users.

Reply:
The if-clause here should protect breaking any website when not running tests: https://github.com/Codeception/c3/blob/f7811b927e9850cb0b922a325ee9b8d95377b54d/c3.php#L34-L36
Also note that any errors encountered could already "break" a website, since a 500 response code is set when no headers are sent yet: https://github.com/Codeception/c3/blob/f7811b927e9850cb0b922a325ee9b8d95377b54d/c3.php#L54

---

I know this part changes behavior when running tests and is perhaps better suited for a new major version. There's actually two points to this change:
- Echo error message. Possibly makes it easier to debug for first-time users. Showing details could be made optional.
- Stop c3 collection when encountering error. Most calls to __c3_error() look like they should stop execution of c3, but do not. For example, when Codeception is not loaded an error is thrown (https://github.com/Codeception/c3/blob/f7811b927e9850cb0b922a325ee9b8d95377b54d/c3.php#L72-L74) but execution continues. This could lead to more errors or unexpected behavior. Another way to solve this is to make those calls throw proper exceptions and catch them to log error and continue normal application execution.